src/BenchmarksApps/SslStream/Server/Program.cs

```diff
@@ -0,0 +1,204 @@
+            e.Cancel = true;
+            ctrlC.Cancel();
+        };
+
+        try
+        {
+            while (!ctrlC.IsCancellationRequested)
+            {
+                var clientSock = await sock.AcceptAsync(ctrlC.Token).ConfigureAwait(false);
+                _ = Task.Run(() => ProcessClient(clientSock, options, sslOptions, ctrlC.Token));
+            }
+        }
+        catch (OperationCanceledException)
+        {
+            // Expected, just return
+        }
+
+        LogMetric("sslstream/error", s_errors);
+        Log("Exitting...");
+    }
+
+    static async Task ReadWriteScenario(SslStream stream, ServerOptions options, CancellationToken cancellationToken)
+    {
+        static async Task WritingTask(SslStream stream, int bufferSize, CancellationToken cancellationToken)
+        {
+            if (bufferSize == 0)
+            {
+                return;
+            }
+
+            var sendBuffer = new byte[bufferSize];
+
+            try
+            {
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    await stream.WriteAsync(sendBuffer, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                // Expected, just return
+            }
+        }
+
+        static async Task ReadingTask(SslStream stream, int bufferSize, CancellationTokenSource cts)
+        {
+            var recvBuffer = new byte[Math.Max(bufferSize, 1)];
+
+            while (true)
+            {
+                var bytesRead = await stream.ReadAsync(recvBuffer, cts.Token).ConfigureAwait(false);
+
+                if (bytesRead > 0 && bufferSize == 0)
+                {
+                    throw new Exception("Client is sending data but the server is not expecting any");
+                }
+
+                if (bytesRead == 0)
+                {
+                    // client closed the connection.
+                    cts.Cancel();
+                    break;
+                }
+            }
+        }
+
+        byte[] recvBuffer = new byte[options.ReceiveBufferSize];
+
+        CancellationTokenSource cts = new CancellationTokenSource();
+        using var reg = cancellationToken.Register(() => cts.Cancel());
+
+        var writeTask = Task.Run(() => WritingTask(stream, options.SendBufferSize, cts.Token));
+        var readTask = Task.Run(() => ReadingTask(stream, options.ReceiveBufferSize, cts));
+
+        await Task.WhenAll(writeTask, readTask).ConfigureAwait(false);
+    }
+
+    static async Task ProcessClient(Socket socket, ServerOptions options, SslServerAuthenticationOptions sslOptions, CancellationToken cancellationToken)
+    {
+        try
+        {
+            using var stream = await EstablishSslStreamAsync(socket, sslOptions, cancellationToken).ConfigureAwait(false);
+
+            if (stream.NegotiatedApplicationProtocol.Protocol.Span.SequenceEqual("Handshake"u8))
```

Comment:
any reason why not to use existing values? There are some optimizations for common values (like h2) and this is going to be somewhat more expensive. I'm not sure if that matters but I feel it would be nice if the benchmark represents common scenario.

Reply:
I wanted to have the server agnostic about what scenario (i.e. only client chooses the scenario and you don't have to restart the server) and didn't want to do the selection by sending an explicit message from the client.

We potentially could assign "h2" and "h3" values to "handshake" and "readwrite" scenarios. to make sure the optimization kicks in